File: hashmap/design-underground-system/design-underground-system.java
class UndergroundSystem {
  Map<Integer, Pair<String, Integer>> checkins;
  Map<String, Pair<Integer, Integer>> checkouts;

  public UndergroundSystem() {
    checkins = new HashMap<>();
    checkouts = new HashMap<>();
  }

  public void checkIn(int id, String stationName, int t) {
    checkins.put(id, new Pair<>(stationName, t));
  }

  public void checkOut(int id, String stationName, int t) {
    Pair<String, Integer> checkin = checkins.get(id);
    String route = checkin.getKey() + "->" + stationName;
    int time = t - checkin.getValue();
    Pair<Integer, Integer> checkout = checkouts.getOrDefault(route, new Pair<>(0, 0));
    checkouts.put(route, new Pair<>(checkout.getKey() + time, checkout.getValue() + 1));
  }

  public double getAverageTime(String startStation, String endStation) {
    String route = startStation + "->" + endStation;
    Pair<Integer, Integer> checkout = checkouts.get(route);
    return (double) checkout.getKey() / checkout.getValue();
  }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t); obj.checkOut(id,stationName,t); double param_3
 * = obj.getAverageTime(startStation,endStation);
 */

File: hashmap/design-underground-system/design-underground-system.js
function UndergroundSystem() {
  this.checkins = {};
  this.checkouts = {};
}

/**
 * @param {number} id
 * @param {string} stationName
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkIn = function (id, stationName, t) {
  this.checkins[id] = [stationName, t];
};

/**
 * @param {number} id
 * @param {string} stationName
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkOut = function (id, stationName, t) {
  const checkin = this.checkins[id];
  const route = checkin[0] + '->' + stationName;
  const time = t - checkin[1];

  if (!this.checkouts[route]) {
    this.checkouts[route] = [0, 0];
  }

  this.checkouts[route][0] += time;
  this.checkouts[route][1]++;
};

/**
 * @param {string} startStation
 * @param {string} endStation
 * @return {number}
 */
UndergroundSystem.prototype.getAverageTime = function (startStation, endStation) {
  const route = startStation + '->' + endStation;
  const checkout = this.checkouts[route];
  return checkout[0] / checkout[1];
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * var obj = new UndergroundSystem()
 * obj.checkIn(id,stationName,t)
 * obj.checkOut(id,stationName,t)
 * var param_3 = obj.getAverageTime(startStation,endStation)
 */
